case-study send function bugs and log format changes

SimonSK opened the pull request:
- [x] send function bug fixes
  - The Send function is removed to avoid confusion. All codes, including tests, should use either SendDEVp2p or SendEthSubProto.
  - Any part of p2p package that previously used `Send` now use `SendDEVp2p`
  - Any part of eth package that previously used `Send` now use `SendEthSubproto`
  - Any part of all other packages that previously used `Send` now use `SendDEVp2p` (as logging is the only difference between the two).
  - `SendItems` should call `SendDEVp2p`. It is used by only p2p. Anywhere that previously used `SendItems` should continue to use it, not the other functions.
- [x] bring all original log messages back
- [x] include peer specific info (id, address, connection type) in every custom log message
  - [x] pass additional information to the send functions
- [x] log to file(s) at `datadir/geth-case-study/logs`. if not set, no logs are written to files, and only the ones with `log-level <= verbosity` are sent to `stderr`
  - `geth-case-study.log`: all log messages with `log-level <= verbosity` go here.
  - `rlpx-received.log`: `RLPXRX` - log level 6
  - `rlpx-sent.log`: `RLPXTX` - log level 7
  - `devp2p-received.log`: `DEVP2PRX` - log level 8
  - `devp2p-sent.log`: `DEVP2PTX` - log level 9
  - `eth-received.log`: `ETHRX` - log level 10
  - `eth-sent.log`: `ETHTX` - log level 11
  - `daofork.log`: `DAOFORK` - log level 12
- [x] add logrotate
  - use case: re-open/create the log files, after `logrotate` occurred (eg log filenames changed).
  - ex) `geth --exec 'admin.logrotate()' attach [/path/to/ipc]`

Log message formats:

- Prefix for all messages (both default and ours):
  `loglevel`|`log timestamp in unix time`
- Dao fork messages:
  `prefix`|`received timestamp`|`support=bool`|`id=nodeid`|`addr=ip:port`|`conn=connection type`
- All other custom messages:
  `prefix`|`sent/received timestamp`|`message type`|`size=int`|`data=...`|`err=...`|`id=nodeid`|`addr=ip:port`[|`conn=connection type`]
  - `RLPX` messages don't have `conn`

SimonSK committed:
p2p.Send completely removed. test cases changed accordingly. a bunch of
them should fail for now due to other unfixed issues.

SimonSK committed:
proper checks inside Send functions.

all test cases, except rlpx_test:TestProtocolHandshake, should pass.

SimonSK committed:
use SendItems where it's needed.

it should pass all tests now.

SimonSK committed:
pass peer id to SendItems

SimonSK committed:
fixes for sending ping/pong messages and correctly reading sent discreason

SimonSK committed:
overall log format changes

SimonSK committed:
logtofile and logrotate

SimonSK committed:
properly prepare logdir

make sure to lock the directory to prevent other processes from writing to the files

SimonSK committed:
devp2p disconnect message is not sent until encryption handshake passes

sid1602 commented:
For the DAO fork messages, I assume the log will contain the nodeid as well?

SimonSK committed:
record received time during p2p handshake

SimonSK commented:
good catch. i totally screwed up the dao fork message. will push the change soon.

SimonSK committed:
daofork messages should have id,addr,conn as well

SimonSK commented on eth/peer.go:
```diff
@@ -234,15 +235,14 @@ func (p *peer) Handshake(network uint64, td *big.Int, head common.Hash, genesis
 	var status statusData // safe to read after two values have been received from errc
-
 	go func() {
 		errc <- p2p.SendEthSubproto(p.rw, StatusMsg, &statusData{
```
`statusData` here is being logged as `struct {}{}`. See `p2p/message.go:116`.

SimonSK replied:
Resolved.

SimonSK commented on p2p/message.go:
```diff
@@ -124,58 +113,63 @@ var dataExcludedMsgs = map[uint64]struct{}{
 }
 
-func SendEthSubproto(w MsgWriter, msgcode uint64, data interface{}, peers ...discover.NodeID) error {
+func SendEthSubproto(w MsgWriter, msgcode uint64, data interface{}, connInfoCtx ...interface{}) error {
```
When sending our status, the data is logged as `struct {}{}`. The size (always 76) is logged correctly, so I think the data passed in is fine. ~~~Not a big concern for now as we don't really care about our status that's sent.~~~ This may affect other sent messages as well, but I don't know if we really care about sent data contents, as long as we are logging the size correctly. (we are omitting all data, except `Status` and `NewBlockHashes` anyways.)

sid1602 replied:
Is this only a logging issue? It seems like it only affects the logging of the data, and not any logging metadata.

SimonSK replied:
Yes. It only affects content of the data.

sid1602 replied:
I think this shouldn't be a problem for us, since our focus is not on the content of the messages, more the messages themselves.

SimonSK replied:
Resolved. It had nothing to do with before/after sending; just a stupid mistake. I'm leaving it as logging after sending, but `time.Now()` will be moved to before.

SimonSK commented on eth/handler.go:
```diff
@@ -321,23 +321,23 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 	}
 	defer msg.Discard()
 
```
@sid1602 Some messages here are logged as size=1. Find out when that happens (especially for `BlockBodiesMsg`, `NodeDataMsg`, and `Receipts`.) Maybe easier to debug at `SendEthSubproto` at `p2p/message.go:116`.

SimonSK commented on p2p/message.go:
```diff
@@ -124,58 +113,63 @@ var dataExcludedMsgs = map[uint64]struct{}{
-	if _, ok := dataExcludedMsgs[msgcode]; ok {
-		data = "<OMITTED>"
+	obj := data
+	if obj, ok = dataExcludedMsgs[msgcode]; ok {
```
@sid1602 Suggestion for debugging the size=1 issue. Try adding another condition when determining whether to omit the data or not. For example, to log data that's less than size 10:
`if obj, ok = dataExcludedMsgs[msgcode]; ok && size > 10 {`

SimonSK commented:
@sid1602 @zzma another logging format issue that's not too concerning.
`ETH_NEW_BLOCK_HASHES` data is being logged like this:
`data=["{\"Hash\":\"0x4a2b1434bdbfae...`
With this PR, certain characters, including backslash, are escaped, so the message now looks like:
`data=["{\\"Hash\\":\\"0x4a2b1434bdbfae...`
The first backslash comes from the following line escaping the double quote:
https://github.com/teamnsrg/go-ethereum/blob/85a1acbfd70a50d81fe6aba073689b6bbf912e06/eth/protocol.go#L151
I think removing the inner `MarshalObj` should fix this problem.
`msgs = append(msgs, blockHash)`

SimonSK commented on p2p/message.go:
```diff
@@ -124,58 +113,63 @@ var dataExcludedMsgs = map[uint64]struct{}{
-	var peer discover.NodeID
-	if len(peers) == 1 {
-		peer = peers[0]
+	err = w.WriteMsg(Msg{Code: msgcode, Size: uint32(size), Payload: r})
```
@sid1602 @zzma I wonder if logging **after** sending the message empties the `data`. That's the only significant change made I see here. If so, then I imagine `>>ETH_NEW_BLOCK_HASHES` will have the same problem. ~~~I cannot verify if other types are affected yet because data for all other types are omitted and the current running instance hasn't sent any `NewBlockHashes` yet. We can comment out the omitting step and verify this.~~~

sid1602 replied:
Couldn't we just log right before we send the message? That should fix this issue, right?

SimonSK replied:
We could, but then we wouldn't know geth actually sent it or failed due to some error.

SimonSK replied:
Confirmed. `>>ETH_NEW_BLOCK_HASHES` messages also show the data as `struct {}{}`. We can either move it to before the message is actually sent, or do a hard copy of `data`.

sid1602 replied:
You said the issue with moving it before the message being sent is we won't know about any errors, right? In that case we should log it after sending, but make a copy of `data`

SimonSK replied:
Resolved. It had nothing to do with before/after sending; just a stupid mistake. I'm leaving it as logging after sending, but `time.Now()` will be moved to before.

SimonSK committed:
when logging sent messages, record time **before** sending messages

SimonSK committed:
fix for incorrect data content in log messages

SimonSK committed:
manually marshaling each block hash is unncessary and does unintended
character-escapes.

SimonSK committed:
refactoring message code to string conversion.
logging unknown eth messages.

SimonSK committed:
the same daofork log message issue fixed in other file.
the function here isn't actually used anywhere but making the change to
keep the codes consistent
